### ISSUE TITLE
Feature request: allow .append() to list instead of direct index assignment

### ISSUE THREAD

**Issue by alexpetralia** [alex-sherman/deco#28]
The following code works:

```
@concurrent
def scrape(link):
    # do work
    pass

x = [None] * 100000 
@synchronized
def populate_df(links):
    for i, link in enumerate(links):
        x[i] = scrape(link)
    return x
```

However, this code does not:

```
@concurrent
def scrape(link):
    # do work
    pass

@synchronized
def populate_df(links):
    x = []
    for i, link in enumerate(links):
        x.append(scrape(link))
    return x
```

Instead it returns a list, one element for each element in the initial `links` list:

```
 <multiprocessing.pool.ApplyResult at 0x7f784ba424a8>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42550>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba425f8>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba426a0>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42748>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba427f0>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42898>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42940>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba429e8>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42a90>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42b38>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42be0>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42c88>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42d30>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42dd8>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42e80>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42f28>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba42fd0>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba480b8>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba48160>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba48208>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba482b0>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba48358>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba48400>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba484a8>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba48550>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba485f8>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba486a0>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba48748>,
 <multiprocessing.pool.ApplyResult at 0x7f784ba487f0>]
```

Would it be possible to get `append` to work?

**Comment by alex-sherman** [comment]
It's awesome that you ask this, it's actually been on my mind lately. There is still a lot to consider, but I think there is a way to make work, maybe I will get around to it as early as this weekend.

If you're interested, I'll describe briefly what's going on. We originally chose to only support index based assignment because it is a very specific case that we thought would be easy for people to refactor their code around. Index based assignment in Python can be transformed to a functionally identical function call to `__setitem__`, making it a subset of this broader category of function calls. Deco currently just defers index based assignment to happen at the next synchronization event, which happens the first time a mutated variable is encountered in a read context. This is where things get complicated, or rather why index based mutation is simpler to handle. Function calls are complicated to reason about because they can mutate extra variables, which could then require a synchronization even before they're read. So now we would need to implement some much more complicated data flow analysis, _or_ make some restrictions on what kind of function calls we allow. Deco was sort of born out of this dichotomy in the first place, it's partially motivated by some earlier work called [Pydron](http://pydron.org/). They designed their system around generality, it has some crazy complicated data flow analysis and it has seen some academic success I think. We designed deco around simplicity and the common case, which allowed us to cut a lot of corners and I think will allow it to eventually become a production ready tool. I think deco will be sort of stuck with this simplicity of generality philosophy, but I also think there may be a way to stay simple and support what you're asking about.

**TL;DR** version is that it's really hard to do this in a general way, but probably there are some reasonable restrictions we can apply and implement something that works for the common cases.

I'm going to leave this issue open until we can make some decision, either on how to appropriately handle deferring function calls, or to not handle them and instead throw an exception.

**Comment by alexpetralia** [comment]
Thank you for the explanation!

**Comment by shantanuo** [comment]
I guess the issue that I face here has to do something with this.

http://stackoverflow.com/questions/41207591/multiprocess-sub-function-does-not-return-any-results

**Comment by alex-sherman** [comment]
This issue will cause problems with that code certainly, but there is another issue which I've responded to in the SO post with some working code.

**Comment by alex-sherman** [comment]
The latest release of `deco` does now include this functionality.
